neuralop/layers/attention_kernel_integral.py

```diff
@@ -0,0 +1,193 @@
+import torch
+from torch import nn
+import math
+from torch.nn.init import xavier_uniform_, zeros_
+
+
+class AttentionKernelIntegral(torch.nn.Module):
+    """
+    Kernel integral transform with attention
+    Computes \int_{Omega} k(x, y) * f(y) dy,
+    where:
+          K(x, y) = \sum_{c=1}^d \q_c(x) * \k_c(y)
+          f(y) = v(y)
+
+    Parameters
+    ----------
+    channels : int, input and output channels
+    heads : int, number of attention heads
+    dim_head : int, dimension of each attention head, determines how many function bases to use for the kernel
+    pos_dim : int, dimension of domain
+    use_pe : bool, whether to use positional encoding
+    project_query : bool, whether to project the query function with pointwise linear layer
+    """
+
+    def __init__(self,
+                 channels,
+                 heads,
+                 dim_head,
+                 pos_dim,
+                 use_pe=True,    # use positional encoding
+                 project_query=True,
+                 ):
+        super().__init__()
+        self.layers = nn.ModuleList([])
+        self.heads = heads
+        self.dim_head = dim_head
+
+        self.project_query = project_query
+        if project_query:
+            self.to_q = nn.Linear(channels, dim_head * heads, bias=False)
+        else:
+            self.to_q = nn.Identity()
+
+        self.to_k = nn.Linear(channels, dim_head * heads, bias=False)
+
+        self.k_norm = nn.InstanceNorm1d(dim_head, affine=False)
+        self.v_norm = nn.InstanceNorm1d(dim_head, affine=False)
+
+        self.to_v = nn.Linear(channels, dim_head * heads, bias=False)
+
+        self.to_out = nn.Linear(dim_head * heads, channels) if dim_head * heads != channels else nn.Identity()
+
+        self.use_pe = use_pe
+        self.pos_dim = pos_dim
+
+        self.init_gain = 1 / math.sqrt(dim_head)
+        self.diagonal_weight = self.init_gain
+        self.initialize_qkv_weights()
+
+    def init_weight(self, weight, inif_fn):
+        # initialization for the projection matrix
+        # basically initialize the weights for each heads as
+        # this requires input dim = dim_head
+        # W = init_fn(W) + I * diagonal_weight
+        # init_fn is usually xavier_uniform_
+        for param in weight.parameters():
+            if param.ndim > 1:
+                dim_head = param.size(0) // self.heads
+                for h in range(self.heads):
+                    inif_fn(param[h * dim_head:(h + 1) * dim_head, :], gain=self.init_gain)
```

Comment:
Something went wrong with formatting here

Reply:
Reformatted this part, let me know if the issues remain